### PR TITLE
Refactor useBillingData to useCheckoutContext and hooking it up to form

### PR DESCRIPTION
**Pull request by senadir**
So this PR changes some things:

1- previously, the returned shipping address didn’t have first and last name in it, this PR adds it.
2- It refactors billingFields to be saved in `useCheckoutContext` reducer, instead of locally.
3- `useBillingData` now exposes six properties:
`email` `setEmail` `billingAddress` `setBillingAddress` `billingData` `setBillingData`.
this is meant to track and update each of them separately, although I think simply `billingData` and `setBillingData` should be enough.
4. It removes the `use shipping as billing` option in the editor since it was useless now.

fixes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1940

### How to test the changes in this Pull Request:

1. Make sure you have some default billing and shipping information in Woo in your profile -> billing
2. in Checkout block, see that those values are hydrated.
3. Update the shipping form
3. toggle off the use shipping as billing option
4. see that billing forms are the same as shipping
5. update shipping now again
6. those changes should not be reflected on billing
7. Toggling and untoggling that option will copy the values.

### CODE REVIEW

File: assets/js/base/context/cart-checkout/checkout/reducer.js [hunk 1]
```diff
@@ -120,6 +121,14 @@ export const reducer = ( state = DEFAULT_STATE, { url, type } ) => {
 				calculatingCount: Math.max( 0, state.calculatingCount - 1 ),
 			};
 			break;
+		case SET_BILLING_DATA:
+			newState = {
+				...state,
+				billingData: {
+					...data,
+				},
+			};
```
[the remark this answers is not in the document]
Comment (senadir) [reply assets/js/base/context/cart-checkout/checkout/reducer.js]: I did assume that redux will do a shallow equal but billing data will invalidate it any, seems like a good plan.

File: assets/js/type-defs/checkout.js [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
+ * @property {function()} clearError                       Dispatches an action
+ *                                                         that clears the hasError
+ *                                                         status for the checkout.
+ * @property {function()} incrementCalculating             Dispatches an action
+ *                                                         that increments the calculating
+ *                                                         state for checkout by one.
+ * @property {function()} decrementCalculating             Dispatches an action
+ *                                                         that decrements the calculating
+ *                                                         state for checkout by one.
+ * @property {function(CartBillingAddress)} setBillingData Dispatches an action
```
[the remark this answers is not in the document]
Comment (senadir) [reply assets/js/type-defs/checkout.js]: it should be fixed now

File: package.json [hunk 3]
```diff
@@ -150,7 +150,7 @@
 		"classnames": "2.2.6",
 		"compare-versions": "3.6.0",
 		"config": "3.3.0",
-		"dinero.js": "1.8.0",
+		"dinero.js": "^1.8.0",
```
[the remark this answers is not in the document]
Comment (senadir) [reply package.json]: yeah I need to reset both of them, guess it lingered from when I reviewed Mike PR

File: assets/js/blocks/cart-checkout/checkout/block.js [hunk 4]
```diff
@@ -23,6 +23,8 @@ import FormattedMonetaryAmount from '@woocommerce/base-components/formatted-mone
 import {
 	CheckoutProvider,
 	useValidationContext,
+	ValidationContextProvider,
```
Comment (nerrad) [inline assets/js/blocks/cart-checkout/checkout/block.js]: This is a rebase conflict resolution error. In Albert's recent pull `ValidationContextProvider` was moved to `blocks/cart-checkout/checkout/frontend.js`

File: assets/js/type-defs/checkout.js [hunk 5]
```diff
@@ -1,28 +1,28 @@
...
+ * @property {function()} clearError                    Dispatches an action
+ *                                                      that clears the hasError
+ *                                                      status for the checkout.
+ * @property {function()} incrementCalculating          Dispatches an action
+ *                                                      that increments the calculating
+ *                                                      state for checkout by one.
+ * @property {function()} decrementCalculating          Dispatches an action
+ *                                                      that decrements the calculating
+ *                                                      state for checkout by one.
+ * @property {function(CartBillingData)} setBillingData Dispatches an action
```
Comment (nerrad) [inline assets/js/type-defs/checkout.js]: While the descriptions _are_ aligned, the variable names are not. they should be aligned as well.  Also when the description alignment means for a narrow column on the 80 column rule, I'll bump up the description to break on the 100 column ruler.

**Pull request merged**